### PR TITLE
IGNITE-12991 Calcite integration. Introduce running query registry & cancellation refactoring

### PR DESCRIPTION
**Pull request by tledkov-gridgain**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/processors/query/QueryEngine.java [hunk 1]
```diff
@@ -36,4 +39,10 @@
      */
     List<FieldsQueryCursor<List<?>>> query(@Nullable QueryContext ctx, String schemaName, String qry, Object... params)
         throws IgniteSQLException;
+
+    /** */
+    Collection<? extends RunningQuery> runningQueries();
```
Comment (korlov42) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/query/QueryEngine.java:44]: If I'm not mistaken, the current approach to cancel query is to get instance of `RunningQuery` and invoke on it a `cancel` method. Unfortunately such an approach doesn't cover the case when current node doesn't run any fragment of the query of interest

Comment (tledkov-gridgain) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/query/QueryEngine.java:44]: I see no problem with it. Looks like the described case is in the scope of `the KILL <query>` command implementation.

Comment (korlov42) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/query/QueryEngine.java:44]: then why do we need a limited API at all?

Comment (tledkov-gridgain) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/query/QueryEngine.java:44]: I guess internal API should represent the modules abilities.
We can introduce the method `QueryEngine#kill(UUID queryId)` when this functionality will be available.

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/Query.java [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+                    throw new IgniteSQLException(
+                        "The query was cancelled",
+                        IgniteQueryErrorCode.QUERY_CANCELED,
+                        new ExecutionCancelledException()
+                    );
+                }
+            }
+        }
+
+        fragments.add(f);
```
Comment (korlov42) [inline modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/Query.java:168]: is it OK if new fragment will be added in `CLOSING` or `CLOSED` state?

File: modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/RootQuery.java [hunk 3]
```diff
@@ -0,0 +1,349 @@
...
+        synchronized (mux) {
+            if (state == QueryState.CLOSED) {
+                throw new IgniteSQLException(
+                    "The query was cancelled while executing.",
+                    IgniteQueryErrorCode.QUERY_CANCELED
+                );
+            }
+
+            if (pctx == null) {
+                state = QueryState.PLANNING;
```
Comment (korlov42) [inline modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/RootQuery.java:267]: I'm wondering, what happens if there will be transition CLOSING --> PLANNING

Comment (tledkov-gridgain) [reply modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/RootQuery.java:267]: Flow is fixed here

**Pull request merged**